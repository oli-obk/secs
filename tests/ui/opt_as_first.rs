use secs::World;

fn optional_components() {
    let world = World::default();

    world.spawn((1_u32,));
    world.spawn((10_u32, "foo"));
    let mut results = vec![];
    world.query::<(&u32, Option<&&str>)>(|_, (i, s)| results.push((*i, s.map(|s| *s))));
    results.sort();
    assert_eq!(&results[..], &[(1, None), (10, Some("foo"))]);

    let mut results = vec![];
    world.query::<(Option<&&str>, &u32)>(|_, (s, i)| results.push((*i, s.map(|s| *s))));
    //~^ ERROR: `Option<&&str>` cannot be the first element of a query
    results.sort();
    assert_eq!(&results[..], &[(10, Some("foo"))]);
}
