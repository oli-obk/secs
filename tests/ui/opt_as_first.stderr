error[E0277]: the trait bound `Option<&&str>: secs::query::Always` is not satisfied
  --> tests/ui/opt_as_first.rs:14:11
   |
14 |     world.query::<(Option<&&str>, &u32)>(|_, (s, i)| results.push((*i, s.map(|s| *s))));
   |           ^^^^^ the trait `secs::query::Always` is not implemented for `Option<&&str>`
   |
   = help: the following other types implement trait `secs::query::Always`:
             &T
             &mut T
   = note: required for `(Option<&&str>, &u32)` to implement `secs::query::Query<'_>`

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0277`.
